src/group_theory/index.lean

```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2021 Thomas Browning. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas Browning
+-/
+
+import group_theory.coset
+import set_theory.cardinal
+
+/-!
+# Index of a Subgroup
+
+In this file we define the index of a subgroup, and prove several divisibility properties.
+
+## Main definitions
+
+- `H.index` : the index of `H : subgroup G`
+
+# Main results
+
+- `index_mul_card` : `H.index * fintype.card H = fintype.card G`
+- `index_dvd_card` : `H.index ∣ fintype.card G`
+- `index_eq_mul_of_le` : If `H ≤ K`, then `H.index = K.index * (H.subgroup_of K).index`
+- `index_dvd_of_le` : If `H ≤ K`, then `K.index ∣ H.index`
+
+-/
+
+namespace subgroup
+
+variables {G : Type*} [group G] (H : subgroup G)
+
+/-- The index of a subgroup as a natural number -/
+noncomputable def index : ℕ :=
+(cardinal.mk (quotient_group.quotient H)).to_nat
+
+lemma index_eq_card [fintype (quotient_group.quotient H)] :
+  H.index = fintype.card (quotient_group.quotient H) :=
+cardinal.mk_to_nat_eq_card
+
+lemma index_mul_card [fintype G] [fintype H] :
+  H.index * fintype.card H = fintype.card G :=
+begin
+  classical,
+  rw H.index_eq_card,
+  convert H.card_eq_card_quotient_mul_card_subgroup.symm,
+end
+
+lemma index_dvd_card [fintype G] : H.index ∣ fintype.card G :=
+begin
+  classical,
+  rw H.index_eq_card,
+  convert H.card_quotient_dvd_card,
+end
+
+variables {H} {K : subgroup G}
+
+/-- If `H ≤ K` then `G/H ≃ G/K × K/H` -/
+noncomputable def quotient_equiv_prod_of_le (h_le : H ≤ K) : quotient_group.quotient H ≃
+  quotient_group.quotient K × quotient_group.quotient (H.subgroup_of K) :=
+{ to_fun := λ a, ⟨quotient_group.mk a.out', quotient_group.mk
+    ⟨(quotient_group.mk a.out' : quotient_group.quotient K).out'⁻¹ * a.out',
+      by rw [←quotient_group.eq', quotient_group.out_eq']⟩⟩,
```

Comment:
Can this be done constructively by using `lift` instead of `out`?

Reply:
I don't think so. I'm pretty sure that the bijection is noncanonical, and really does depend on the choice of of the lift.

Reply:
For instance, if I start with `refine λ a, (quot.factor _ _ (λ _ _, @@h_le) a, _),` then I have the first half of the prod computably.

Reply:
The bijection is (G/H) = (G/K) x (K/H).
The first factor (G/H) -> (G/K) is computable and canonical.
The second factor (G/H) -> (K/H) is noncanonical, as it requires choosing a representative from each coset of K (and then you define the map by translating by the corresponding representative).

Reply:
I suspect defining just the first factor computably may still give a nicer proof

Reply:
That's an interesting idea. I'll look into it later today.

Reply:
You might also be able to do the whole thing computably if you take a right_inverse of quotient.mk as an argument. You'll then be able to add a special case where the right-inverse is just out'.

Reply:
Good point. I'll give it a try.

Reply:
@eric-wieser I've added a computable version.